invoke-ai/InvokeAI
# Installer rewrite in Python

## Summary

This PR rewrites the core of the installer in Python for cross-platform compatibility. Filesystem path manipulation, platform/arch decisions and various edge cases are handled in a more convenient fashion. The original `install.bat.in`/`install.sh.in` scripts are kept as entrypoints for their respective OSs, but only serve as thin wrappers to the Python module.

In addition, it:

- builds and **packages the .whl with the installer**, so that downloading a versioned installer will guarantee installation of the same version of the application.
- updates shell entrypoints: 
  - new commands are `invokeai`, `invokeai-configure`, `invokeai-ti`, `invokeai-merge`.
  - these commands will be available in the activated `.venv` or via the launch scripts
  - `invoke.py` and `configure_invokeai.py` scripts are deprecated but kept around for backwards compatibility and keeping users' surprise to a minimum.
- introduces a new `ldm/invoke/config` package and moves the `configure_invokeai` script into it. Similarly, movers Textual Inversion script and TUI to `ldm/invoke/training`.
- moves the `configs` directory into the `ldm/invoke/config` package for easy distribution.
- updates documentation to reflect all of the above changes
- fixes a failing test
- reduces wheel size to 3MB (from 27MB) by excluding unnecessary image files under `assets`

:warning: self-updating functionality and ability to install arbitrary versions are still WIP. For now we can recommend downloading and running the installer for a specific version as desired.

## Testing the source install

From the cloned source, check out this branch, and:

`$ python3 installer/main.py --root <path_to_destination>`

Also try:

`$ python3 installer/main.py ` - will prompt for paths
`$ python3 installer/main.py --yes` - will not prompt for any input

- try to combine the `--yes` and `--root` options
- try to install in destinations with "quirky" paths, such as paths containing spaces in the directory name, etc.

## Testing the packaged install ("Automated Installer"):

Download the [InvokeAI-installer-v2.3.0+a0.zip](https://github.com/invoke-ai/InvokeAI/files/10533913/InvokeAI-installer-v2.3.0%2Ba0.zip) file, unzip it, and run the install script for your platform (preferably in a terminal window)

OR make your own: from the cloned source, check out this branch, and:

```
cd installer
./create_installer.sh
# (do NOT tag/push when prompted! just say "no")
```

This will create the installation media: `InvokeAI-installer-v2.3.0+a0.zip`. The installer is now *platform-agnostic* - meaning, both Windows and *nix install resources are packaged together. 

Copy it somewhere as if it had been downloaded from the internet. Unzip the file, enter the created `InvokeAI-Installer` directory, and run `install.sh` or `install.bat` as applicable your platform.

:warning: NOTE!!! `install.sh` accepts the same arguments as are applicable to the Python script, i.e. you can `install.sh --yes --root ....`. This is NOT yet supported by the Windows `.bat` script. Only interactive installation is supported on Windows. (this is still a TODO).
